src/main/java/net/sf/mpxj/ManuallyScheduledTaskCalendar.java

```diff
@@ -0,0 +1,163 @@
+ */
+public class ManuallyScheduledTaskCalendar extends ProjectCalendar
+{
+   public ManuallyScheduledTaskCalendar(ProjectCalendar calendar, ResourceAssignment assignment)
+   {
+      super(calendar.getParentFile(), true);
+      m_calendar = calendar;
+      m_assignment = assignment;
+      m_assignment_start_date = m_assignment.getStart().toLocalDate();
+      m_assignment_end_date = m_assignment.getFinish().toLocalDate();
+   }
+
+   /**
+    * Retrieves the working hours on the given date.
+    *
+    * @param date required date
+    * @return working hours
+    */
+   @Override protected ProjectCalendarHours getRanges(LocalDate date)
+   {
+      ProjectCalendarHours effectiveRanges = m_calendar.getRanges(date);
+      // If today is not a working day then find first ProjectCalendarRange with working time.
+      if (effectiveRanges.isEmpty() && (date.equals(m_assignment_start_date) || date.equals(m_assignment_end_date)))
+      {
+         // Date is not a working day.
+         // Find first ProjectCalendarRange with working time. Starting on Tuesday(!).
+         // Using [Default] calendar - ignoring exceptions and work week rules. Uses this a basis for all calculations.
+         if (m_calendar.getDayType(DayOfWeek.TUESDAY) == DayType.WORKING)
+         {
+            effectiveRanges = m_calendar.getHours(DayOfWeek.TUESDAY);
+         }
+         else
+            if (m_calendar.getDayType(DayOfWeek.WEDNESDAY) == DayType.WORKING)
+            {
+               effectiveRanges = m_calendar.getHours(DayOfWeek.WEDNESDAY);
+            }
+            else
+               if (m_calendar.getDayType(DayOfWeek.THURSDAY) == DayType.WORKING)
+               {
+                  effectiveRanges = m_calendar.getHours(DayOfWeek.THURSDAY);
+               }
+               else
+                  if (m_calendar.getDayType(DayOfWeek.FRIDAY) == DayType.WORKING)
+                  {
+                     effectiveRanges = m_calendar.getHours(DayOfWeek.FRIDAY);
+                  }
+                  else
+                     if (m_calendar.getDayType(DayOfWeek.SATURDAY) == DayType.WORKING)
+                     {
+                        effectiveRanges = m_calendar.getHours(DayOfWeek.SATURDAY);
+                     }
+                     else
+                        if (m_calendar.getDayType(DayOfWeek.SUNDAY) == DayType.WORKING)
+                        {
+                           effectiveRanges = m_calendar.getHours(DayOfWeek.SUNDAY);
+                        }
+                        else
+                           if (m_calendar.getDayType(DayOfWeek.MONDAY) == DayType.WORKING)
+                           {
+                              effectiveRanges = m_calendar.getHours(DayOfWeek.MONDAY);
+                           }
+         // In case the calendar has no working days. Normally Project blocks the creation of such calendar.
+         if (effectiveRanges.isEmpty())
```

Comment:
Can we move this out if the `if` statement block to just before the `if (date.equals(m_assignment_start_date))` line to catch any possible case where we have an empty range.

Reply:
I moved it. But it was impossible due to the check in line 53.

Reply:
I am happy with the solution and it covers all scenarios I have identified so far.
In case I find more I will open a new Issue / PR.

So happy to complete the PR.